### ISSUE TITLE
ResourceExhaustedError (see above for traceback): OOM when allocating tensor

### ISSUE THREAD

**Issue by nadianaji** [tensorflow/tensorflow#23198]
Hi, 
I want to feed CNN with my database. for this aim, I use ImageDataGenerator and I put my code here. but after implementing it produces this error:  
"ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[32,32,512,512] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[Node: conv2d_1/convolution = Conv2D[T=DT_FLOAT, data_format="NCHW", dilations=[1, 1, 1, 1], padding="SAME", strides=[1, 1, 1, 1], use_cudnn_on_gpu=true, _device="/job:localhost/replica:0/task:0/device:GPU:0"](conv2d_1/convolution-0-TransposeNHWCToNCHW-LayoutOptimizer, conv2d_1/kernel/read)]]"

my database has 3 folders( train, test and validation). train and validation have 7 classes. train contains 564 images and validation contains 191 images. it is smal database but I do not know why it produces this error! could you please guide me about this problem? I do not know how can I solve this?

```
import numpy as np
import matplotlib.pyplot as plt
from keras.preprocessing.image import ImageDataGenerator
from sklearn.utils import shuffle
from sklearn.cross_validation import train_test_split
from keras import backend as K
#K.set_image_dim_ordering('th')

from keras.utils import np_utils
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, Flatten
from keras.layers.convolutional import Conv2D, MaxPooling2D
from keras.optimizers import SGD,RMSprop,adam


train_datagen = ImageDataGenerator(
        rescale=1./255,
        shear_range=0.2,
        zoom_range=0.2,
        horizontal_flip=True)
#
valid_datagen = ImageDataGenerator(
        rescale=1./255,
        shear_range=0.2,
        zoom_range=0.2,
        horizontal_flip=True)
#
test_datagen = ImageDataGenerator(rescale=1./255)
#
train_generator = train_datagen.flow_from_directory(
    directory=r"E:\databasetest\train",
    target_size=(512, 512),
    color_mode="grayscale",
    batch_size=32,
    class_mode="categorical",
    shuffle=True,
    seed=42
)
#
valid_generator = valid_datagen.flow_from_directory(
   directory=r"E:\databasetest\validation",
    target_size=(512, 512),
    color_mode="grayscale",
    batch_size=32,
    class_mode="categorical",
    shuffle=True,
    seed=42
)
#
test_generator = test_datagen.flow_from_directory(
    directory=r"E:\databasetest\test",
    target_size=(512, 512),
    color_mode="grayscale",
    batch_size=16,
    class_mode=None,
    shuffle=False,
    seed=42
)
#
## neural network model
model = Sequential()
model.add(Conv2D(32, (3,3),border_mode='same', input_shape = (512, 512, 1), activation = 'relu'))
model.add(Activation('relu'))
model.add(Conv2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.5))

model.add(Conv2D(64, 3, 3))
model.add(Activation('relu'))
#model.add(Convolution2D(64, 3, 3))
#model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.5))

model.add(Flatten())
model.add(Dense(64))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(7))
model.add(Activation('softmax'))

model.summary()

model.compile(loss = 'categorical_crossentropy',
              optimizer = 'rmsprop',
              metrics = ['accuracy'])
STEP_SIZE_TRAIN=train_generator.n//train_generator.batch_size
STEP_SIZE_VALID=valid_generator.n//valid_generator.batch_size
model.fit_generator(generator=train_generator,
                    steps_per_epoch=STEP_SIZE_TRAIN,
                    validation_data=valid_generator,
                    validation_steps=STEP_SIZE_VALID,
                    epochs=10
)
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by ChristianLagares** [comment]
A `ResourceExhaustedError` is typically raised when a particular tensor could not be allocated on the requested device. 
@nadianaji could you add additional details on your GPU? Is the same GPU being used to drive a display? What's the output of `nvidia-smi`?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by ChristianLagares** [comment]
@nadianaji a 1050 Ti has a limited amount of memory (4 GB). My recommendations:
    - Try reducing your `target_size` from `(512, 512)` to something more manageable. I would start from a small patch such as 64 or 128 and move from there until the `ResourceExhaustedError` is raised again.
    - Try reducing the batch size (this wouldn't be my first choice, but something like 16 could also help reduce pressure on the memory).

I am almost certain that reducing your `target_size` would definitely help. Let me know if this helps.

**Comment by nadianaji** [comment]
thank you for the response. what is this memory (4GB)? is it different from RAM?

**Comment by ChristianLagares** [comment]
You are most welcomed. 

Typically discrete GPUs have dedicated graphics memory. If I’m not mistaken, the 1050 Ti has 4 GB of GDDR5 memory. Although you have 16 GB of main memory for your host CPU, your GPU is limited to the available discrete memory. Since you are also driving a display, you probably have around 3.5 GB available for TensorFlow.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
@nadianaji Is this still an issue for you?

**Comment by ymodak** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by brendavsantos** [comment]
Thank you! I actually found some errors but now they are all fixed and I am getting a new error:
![image](https://user-images.githubusercontent.com/48032705/62428973-bd833b80-b6de-11e9-84e0-70d7e6168758.png)

**Comment by aluamine** [comment]
Hello!

I stumbled on this post, and I know that it has been closed for more than a year now. However, I got the same problem when using 3D CNN with Keras on my machine.
The weird thing is that when checking the memory allocation on my machine it does not even get close to saturation before this OOM error appears.
I have 64 GB of RAM, 2 GPUs (NVIDIA Quadro P5000). The GPUs have a dedicated memory of 16 GB and 32 GB of shared memory.
I'm only using 2 layers of conv3D of size 16 each. When I go up to 32, I get the OOM error.

Can you please help? Is this purely a memory hardware problem? Or could it be something else? (software, tensorflow, etc.)

Thank you!

Here is the code for building the model:

    model = Sequential()
    model.add(Conv3D(32, kernel_size=(3,3,3),  kernel_regularizer=l2(l2_lambda), activation='selu', padding='same', kernel_initializer='glorot_uniform', input_shape=(64, 64, 38, 1)))
    model.add(MaxPooling3D(pool_size=(2,2,2)))

    model.add(Conv3D(32, kernel_size=(3,3,3),  kernel_regularizer=l2(l2_lambda), activation='selu', padding='same', kernel_initializer='glorot_uniform', input_shape=(64, 64, 38, 1)))
    model.add(MaxPooling3D(pool_size=(2,2,2)))

    model.add(Flatten())
    model.add(Dense(256, kernel_regularizer=l2(l2_lambda), activation='selu', kernel_initializer='glorot_uniform'))

    model.add(Dense(no_classes, activation='softmax'))
    model = multi_gpu_model(model, gpus = 2)
    # Compile the model
    model.compile(loss=keras.losses.categorical_crossentropy,
                optimizer=keras.optimizers.Adam(lr=learning_rate),
                metrics=['accuracy'])
